<a href="https://colab.research.google.com/github/jaleftwi/MAT421_Modules/blob/main/MAT421_ModuleC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Module C**

---
Moving on to the next section in MAT 421, Module C will cover the following topics:

*   19.1 Root Finding Problem Statement
*   19.2 Tolerance
*   19.3 Bisection Method
*   19.4 Newton-Raphson Method
*   19.5 Root Finding in Python

## **Section 19.1 Root Finding Problem Statement**
---
As a basic foundational knowledge- the 'root' or oftentimes the 'zero' is the value x\* of a function f(x) such that f(x\*) = 0. Roots are one of the most important properties of a function. As many engineers and mathematicians are well aware- root finding is essential in the processes of optimization and problem solving.

Of course, the difficulty of root finding can vary from function to function. While f(x) = x - 10 can easily be solved for x = 10 when f(x) = 0, and though techniques to find precise roots exist for quadratic polynomial functions such as the 'quadratic formula'; most functions do not possess a convenient analytic method to obtain a precise root.

For this purpose, several methods to approximate roots have been formulated to best identify a close solution. While there are be functions that do not have roots absolutely, many of these methods are adept in finding roots in even the most complicated functions in which roots do indeed exist.

## **Section 19.2 Toleranace**
---
One thing to touch upon before delving into these methods is the idea of 'tolerance'. Since these methods are not purely 'analytic' in that they do not always determine exact solutions; they do however produce approximations of varying degrees of accuracy that are 'close enough' to the true solution.

As consequence of these approximations, some deviation or error is bound to occur. For this reason, in using these methods we must carefully consider the tolerance we wish to accept. Tolerance is in this defintion: the level of error that is found acceptable within the scope of the problem.

A solution is considered acceptable when the error produced between the approximate and true solutions are within the set level of tolerance. In this case, |f(x*)| < tolerance is the 'check' used to determine validity. However, other methods opt to use the deviation between two approximate solutions as the means to prove validity. In this case, |xi+1 - xi| < tolerance is represented with 'i' signifying a single iteration of the process used to deteremine the root.

## **Section 19.3 Bisection Method**
---
At its basis, the bisection method utilizes a concept known as the intermediete value theorem. According to the theorem: if f(x) is a continuous function and there exists a,b (a < b) such that f(a) and f(b) are opposite in sign (one is positive and one is negative), then there exists a 'c' such that a < c < b  in which f(c) = 0. Therefore, somewhere between a and b exists a value which would be a root of the function.

Through an iterative process of narrowing down the range of 'a' and 'b', the approximate midpoint m = (a+b)/2 approaches within the range of tolerance that we set for the 'true' solution. The initial midpoint will often not be suitable as the approximate soltuion, in which case the value m will either replace a (f(a) > 0 and f(m) > 0) as the 'upper bound' of the range or replace b (f(b) < 0 and f(b) < 0) as the 'lower bound' of the range. Once replaced, the midpoint is taken again, and the process continues until tolerance is reached.

The code below is a demonstration of the process in Python:

In [7]:
import numpy as np

def my_bisection(f, a, b, tolerance):
  # This function takes a mathematical function f to approximate a root
  # between a and b until tolerance is reached. In this case
  # this is when |f(m)| < tolerance with m being the midpoint between a and b

  # First check if a and b are opposite signs
  if np.sign(a) == np.sign(b):
    raise Exception(
        "There is not a root between a and b"
    )
  
  # Get midpoint
  m = (a+b) / 2

  # Check if midpoint is within tolerance, if not replace a bound:
    # Is m within range of the solution?
  if np.abs(f(m)) < tolerance:
    return m

    # Is m an improvement on the a bound?
  elif np.sign(f(a)) == np.sign(f(m)):
    return my_bisection(f, m, b, tolerance)   # Perform the process again using a = m

    # Is m an improvement on the b bound?
  elif np.sign(f(b)) == np.sign(f(m)):
    return my_bisection(f, a, m, tolerance)   # Perform the process again using b = m

  # The process recursively repeates until f(m) < tolerance

In [8]:
# Demostration
f = lambda x: x**2 - 100      # y = x^2 - 100, roots: -10 and 10

root = my_bisection(f, 0, 20, 8)
print("Root =", root)

Root = 10.0


## **Section 19.4 Newton-Raphson Method**
---
Sometimes the midpoint is not the most efficient value to use to approximate the root. For this reason, other iterative methods have been used to approximate the root- including the Newton-Raphson Method.

The Newton-Raphson methods utilizes the derivative of the function at a chosen point x0, and uses that derivative (the slope at that point) to create a 'linear approximation' of the root. This iterative appoaches takes an initial guess x0, and approximates f = f(x0) + f'(x0)(x1 - x0) +/- tolerance. When solving for the next step, x1 = x0 - f(x0)/f'(x0).

As the iterations continued, the guessed next approximation becomes more and more accurate, each using the previous guess as its 'initial guess'.

The process can also be expedited into Python, as shown below:

In [11]:
import numpy as np

f = lambda x: x**2 - 36     # y = x^2 - 36  roots: x = -6 and x = 6
fprime = lambda x : 2*x     # Derivative must be written directly, y' = 2x
xValue = 4.5                # Intial Guess of the root

approximation = xValue - (f(xValue))/(fprime(xValue))
print("Approximation =", approximation)

Approximation = 6.25


This process is the repeated iteratively with the approximation replacing the initial 'xValue'. In general, the Newton-Raphson Method is quicker than the Bisection Method.

**Homework Question**: *Find a function f(x) and guess for the root of f,x0, such that the Newton-Raphson method would oscillate between x0 and −x0 indefinitely.*

**Ans:** In this case, -x0 = x0 - (f(x0)/f'(x0)) and x0 = -x0 - (f(-x0)/f'(-x0))

Because of this, any solution in which x0 = (1/2) * (f(x0) / ((f'(x0))) is appropiate as a solution. In a quadratic a^2*x + b*x + c, this is achieved when c = 3*a*x^2 + b*x.

Thus, if **f(x) = x^2 + x + 80**, and the initial guess **x0 = 5**, then it will oscillate from x0 = 5 to x1 = -5.

## **Section 19.5 Root Finding in Python**
---
Since root finding is a common problem persisting across diverse fields, it should not come as a surprise that the programming language Python comes well equipped with functions that make finding roots much easier. In these cases, we will discuss the Python function '*f_solve*' from *scipy.optimize*.

While not all of the arguments that can be used for '*f_solve*' will be specified here, the documentation for the function online can be used as a supplement to fully understand its implication. That said, there are two aregument that will be touched on breifly here: the function and the intial guess. We will show that implementation below: 



In [6]:
from scipy.optimize import fsolve

f = lambda x: x**3 + 30*x**2 - 100*x + 20

fsolve(f, 20)

array([2.83072253])

From the intial guess, '*f_solve*' detemined **one** approximate soltution to the function as x = 2.8307...

**Note:** Python also accepts multiple initial guesses formed as an array to determine multiple approximate solutions.

In [5]:
from scipy.optimize import fsolve

f = lambda x: x**3 + 30*x**2 - 100*x + 20

fsolve(f, [20, -10])

array([2.83072253, 0.21381248])